_bleio/adapter_.py

```diff
@@ -0,0 +1,222 @@
+        # This is a concurrent.Future.
+        future = asyncio.run_coroutine_threadsafe(coro, self._bleak_loop)
+        return future.result(timeout)
+
+    @property
+    def enabled(self) -> bool:
+        return self._enabled
+
+    @enabled.setter
+    def enabled(self, value: bool):
+        self._enabled = value
+
+    @property
+    def address(self) -> Address:
+        # bleak has no API for the address yet.
+        return None
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @name.setter
+    def name(self, value: str):
+        self._name = value
+
+    def start_advertising(
+        self,
+        data: Buf,
+        *,
+        scan_response: Buf = None,
+        connectable: bool = True,
+        interval: float = 0.1
+    ) -> None:
+        raise NotImplementedError("Advertising not implemented")
+
+    def stop_advertising(self) -> None:
+        """Stop sending advertising packets."""
+        raise NotImplementedError("Advertising not implemented")
+
+    def start_scan(
+        self,
+        prefixes: Buf = b"",
+        *,
+        buffer_size: int = 512,  # pylint: disable=unused-argument
+        extended: bool = False,  # pylint: disable=unused-argument
+        timeout: float = None,
+        interval: float = 0.1,  # pylint: disable=unused-argument
+        window: float = 0.1,  # pylint: disable=unused-argument
+        minimum_rssi: int = -80,
+        active: bool = True  # pylint: disable=unused-argument
+    ) -> Iterable:
+        """
+        Starts a BLE scan and returns an iterator of results. Advertisements and scan responses are
+        filtered and returned separately.
+
+        :param sequence prefixes: Sequence of byte string prefixes to filter advertising packets
+            with. A packet without an advertising structure that matches one of the prefixes is
+            ignored. Format is one byte for length (n) and n bytes of prefix and can be repeated.
+        :param int buffer_size: the maximum number of advertising bytes to buffer.
+        :param bool extended: When True, support extended advertising packets.
+           Increasing buffer_size is recommended when this is set.
+        :param float timeout: the scan timeout in seconds.
+          If None, will scan until `stop_scan` is called.
+        :param float interval: the interval (in seconds) between the start
+           of two consecutive scan windows.  Must be in the range 0.0025 - 40.959375 seconds.
+        :param float window: the duration (in seconds) to scan a single BLE channel.
+           window must be <= interval.
+        :param int minimum_rssi: the minimum rssi of entries to return.
+        :param bool active: retrieve scan responses for scannable advertisements.
+        :returns: an iterable of `_bleio.ScanEntry` objects
+        :rtype: iterable"""
+
+        scanner = BleakScanner(loop=self._bleak_loop)
```

Comment:
Yes, bleak uses bluez. I saw that you used `hcitool`, but `hcitool` is part of the bluez source tree, so is it not also using bluez? Also I've read that it's deprecated. If there's filtering going on, it might be some defaults that bleak set on bluez.

What filtering specifically did you see? Could we get around it by doing very short scanning intervals, if the filtering is during one scanning period? Already I am using 0.25 second scanning periods to simulate real-time scanning results.

Reply:
`hcitool` has raw hci access to the ble module. When I did this, I dug all the way down and found that the linux kernel always deduplicates scans by address iirc. There is no setting otherwise. https://marc.info/?l=linux-bluetooth&m=158225950522806&w=2

Whether 0.25 seconds is enough depends on the configuration of the broadcaster. I was testing this with the `missed-message-count` of the BroadcastNet bridge. It tracks how many sequence numbers are missed from the broadcasts and is one way to get a feel for how many are missed.

Reply:
I'll open an issue about this, for sure, to be investigated and fixed. I'd also like to lobby for them to fix that thing in bluez, since one should be able to turn off the deduplication. I don't want to defer releasing the library until fixing it, though, since it has a bunch of use cases right now that aren't advertising-only.

Reply:
I don't know that the BLE adapter is always accessible via hcitool, though it probably usually is. I don't know if it supports any other adapters besides via HCI.

Reply:
I'm pretty sure all BLE adapters are HCI over one transport or another. Is it a ton of work to use hcitool for scanning on Linux? I'm a bit frustrated that I did a bunch of work on this and it's now being replaced by code that won't work as well.

Reply:
I'm sorry, I can just use your code. I didn't see the conversation you had on the bluez email list until just now; I was only looking at the kernel flag.

Whether `hcitool` and other tools were deprecated and removed or not is really confusing. The Arch Linux people say they were, (as of bluez 5.44?), but they are still in the bluez.org download packages. So I'll include your code but fallback to the current implementation.

Reply:
Just to be clear, I am not using ArchLinux; that's just where I found the discussion. Ubuntu 20.04 uses bluez 5.43.

Reply:
Falling back would be awesome. My main platform of concern is Raspbian.

Reply:
I added and adapted your original code. I added a check for `hcitool` having the correct permissions by attempting to do an `hcitool cmd <noop>`, which will fail without the permissions.